fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java

```diff
@@ -634,4 +641,122 @@ public String getMaterializedViewDdlStmt(boolean simple) {
+                }
+            }
+        } else if (partitionInfo instanceof ExpressionRangePartitionInfo) {
+            // partitions to refresh
+            // 1. dropped partitions
+            // 2. newly added partitions
+            // 3. partitions loaded with new data
+            return getPartitionNamesToRefreshForPartitionBy();
+        } else {
+            throw UnsupportedException.unsupportedException("unsupported partition info type:"
+                    + partitionInfo.getClass().getName());
+        }
+        return Sets.newHashSet();
+    }
+
+    private Set<String> getPartitionNamesToRefreshForPartitionBy() {
+        Expr partitionExpr = getPartitionRefTableExprs().get(0);
+        Pair<Table, Column> partitionInfo = getPartitionTableAndColumn();
+        // if non-partition-by table has changed, should refresh all mv partitions
+        Table partitionTable = partitionInfo.first;
+        for (BaseTableInfo tableInfo : baseTableInfos) {
+            Table table = tableInfo.getTable();
+            if (!table.isLocalTable()) {
+                return Sets.newHashSet();
+            }
+            if (table.getTableIdentifier().equals(partitionTable.getTableIdentifier())) {
+                continue;
+            }
+            Set<String> partitionNames = getNeedRefreshPartitionNames(table);
+            if (!partitionNames.isEmpty()) {
+                return getPartitionNames();
+            }
+        }
+        // from here, partitionTable must be OlapTable or MaterializedView
+        OlapTable basePartitionTable = (OlapTable) partitionTable;
+
+        // check partition-by table
+        Set<String> needRefreshMvPartitionNames = Sets.newHashSet();
+
+        Map<String, Range<PartitionKey>> basePartitionMap = basePartitionTable.getRangePartitionMap();
+        Map<String, Range<PartitionKey>> mvPartitionMap = getRangePartitionMap();
+        PartitionDiff partitionDiff = getPartitionDiff(partitionExpr, partitionInfo.second,
+                basePartitionMap, mvPartitionMap);
+        needRefreshMvPartitionNames.addAll(partitionDiff.getDeletes().keySet());
+        for (String deleted : partitionDiff.getDeletes().keySet()) {
+            mvPartitionMap.remove(deleted);
+        }
+        needRefreshMvPartitionNames.addAll(partitionDiff.getAdds().keySet());
+        for (Map.Entry<String, Range<PartitionKey>> addEntry : partitionDiff.getAdds().entrySet()) {
+            mvPartitionMap.put(addEntry.getKey(), addEntry.getValue());
+        }
+
+        Map<String, Set<String>> baseToMvNameRef = SyncPartitionUtils
+                .generatePartitionRefMap(basePartitionMap, mvPartitionMap);
+        Map<String, Set<String>> mvToBaseNameRef = SyncPartitionUtils
+                .generatePartitionRefMap(mvPartitionMap, basePartitionMap);
+
+        if (partitionExpr instanceof SlotRef) {
+            Set<String> baseChangedPartitionNames = getNeedRefreshPartitionNames(partitionTable);
```

Comment:
This function call could put it outside the if struct

Reply:
done